### PR TITLE
Vf damage fix

### PR DESCRIPTION
**Pull request by perkinslr**
## Additions

- projectiles leaving vehicles spawn fragments
- mod option to make all components reachable 

## Changes
- vehicles die when all reachable components are destroyed
 
## Reasoning

Currently, some vehicles are nearly unkillable, when they have an external component that can never be hit due to not being on the vehicle perimiter.  This resolves that in either of 2 ways, depending on player setting.  Both have edge cases involving the AI.

## Alternatives

Force one choice.  
Let some vehicles remain effectively immortal.

## Testing

Check tests you have performed:
- [x] Compiles without warnings
- [ ] Game runs without errors
- [ ] (For compatibility patches) ...with and without patched mod loaded
- [ ] Playtested a colony (specify how long)

### CODE REVIEW

File: Source/CombatExtended/CombatExtended/ModSettings/Settings.cs [hunk 1]
```diff
@@ -188,6 +188,8 @@ public class Settings : ModSettings, ISettingsCE
 
     #region Compatibility Modsettings
     public bool patchArmorDamage = true;
+    public bool hitRandomVehicleComponents = false;
```
Comment (ViralReaction) [inline Source/CombatExtended/CombatExtended/ModSettings/Settings.cs]: Should probably be set to true as default

File: Source/VehiclesCompat/VehiclesCompat/VehicleSettings.cs [hunk 2]
```diff
@@ -21,6 +21,16 @@ public void DoWindowContents(Listing_Standard list)
...
+            list.CheckboxLabeled("CE_Settings_FragmentsFromVehicles_Title".Translate(), ref Controller.settings.fragmentsFromVehicles, "CE_Settings_FragmentsFromVehicles_Desc".Translate());
+        }
+        else
+        {
+            Controller.settings.fragmentsFromVehicles = false;
+        }
     }
-
+#endif
 }
```
Comment (ViralReaction) [inline Source/VehiclesCompat/VehiclesCompat/VehicleSettings.cs]: Should be 
`#endif
    }
}` instead of `} #endif }` otherwise causes non-Debug builds not to compile.

File: Source/VehiclesCompat/VehiclesCompat/VehicleArmorPatch.cs [hunk 3]
```diff
@@ -207,11 +264,60 @@ public static void ApplyDamageCE(this VehicleStatHandler stats, ref DamageInfo d
...
+        var baseVel = frag.velocity;
+        frag.velocity *= Mathf.Sqrt(keRatio);
+        frag.initialSpeed = frag.shotSpeed = frag.velocity.magnitude * GenTicks.TicksPerRealSecond;
+        frag.ExactPosition = new Vector3(cell.x, height, cell.z);
+        report?.AppendLine($"Spawning fragment");
+        report?.AppendLine($"angle: {frontArc}");
+        report?.AppendLine($"damage: {dinfo.Amount} / {baseAmount}");
+        report?.AppendLine($"mass: {frag.mass} / {baseMass}");
+        report?.AppendLine($"keRatio: {keRatio}");
+        report?.AppendLine($"velocity {frag.velocity} / {baseVel}");
```
Comment (ViralReaction) [inline Source/VehiclesCompat/VehiclesCompat/VehicleArmorPatch.cs]: Should be a single null check on report and than early return. Checked and compiler does not optimize the repeated null checks.

**Review by ViralReaction — CHANGES_REQUESTED** [review]

File: Source/VehiclesCompat/VehiclesCompat/VehicleArmorPatch.cs (hunk 3 above)
Comment (perkinslr) [reply Source/VehiclesCompat/VehiclesCompat/VehicleArmorPatch.cs]: That's *probably* handled by the JIT, but there's no harm in helping it.  At the least it'll reduce the bytecode size by a bit.

File: Source/VehiclesCompat/VehiclesCompat/VehicleSettings.cs (hunk 2 above)
Comment (perkinslr) [reply Source/VehiclesCompat/VehiclesCompat/VehicleSettings.cs]: Should probably make the github builder compile in both release modes, just to catch that sort of thing.  Thanks!

**Review by ViralReaction — DISMISSED** [review]

**Review by ViralReaction — DISMISSED** [review]

**Review by N7Huntsman — APPROVED** [review]
